// @motorcycle/dom/lib/fromEvent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Stream = require('most/lib/Stream');

var _Stream2 = _interopRequireDefault(_Stream);

var _MulticastSource = require('most/lib/source/MulticastSource');

var _MulticastSource2 = _interopRequireDefault(_MulticastSource);

var _forEach = require('fast.js/array/forEach');

var _forEach2 = _interopRequireDefault(_forEach);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var tryEvent = function tryEvent(t, x, sink) {
  try {
    sink.event(t, x);
  } catch (e) {
    sink.error(t, e);
  }
};

var EventAdapter = function EventAdapter( // eslint-disable-line
init, event, source, useCapture, sink, scheduler) {
  this.event = event;
  this.source = source;
  this.useCapture = useCapture;

  var addEvent = function addEvent(ev) {
    tryEvent(scheduler.now(), ev, sink);
  };

  this._dispose = init(source, event, addEvent, useCapture);
};

EventAdapter.prototype.dispose = function dispose() {
  return this._dispose(this.event, this.source);
};

var initEventTarget = function initEventTarget(source, event, addEvent, useCapture) {
  // eslint-disable-line
  (0, _forEach2.default)(source, function (s) {
    return s.addEventListener(event, addEvent, useCapture);
  });

  var dispose = function dispose(_event, target) {
    (0, _forEach2.default)(target, function (t) {
      return t.removeEventListener(_event, addEvent, useCapture);
    });
  };

  return dispose;
};

function EventTargetSource(event, source, useCapture) {
  this.event = event;
  this.source = source;
  this.useCapture = useCapture;
}

EventTargetSource.prototype.run = function run(sink, scheduler) {
  return new EventAdapter(initEventTarget, this.event, this.source, this.useCapture, sink, scheduler);
};

var fromEvent = function fromEvent(event, source) {
  var useCapture = arguments.length <= 2 || arguments[2] === undefined ? false : arguments[2];

  // is not a NodeList
  if (!source.length) {
    throw new Error('source must be a NodeList or an Array of DOM Nodes');
  }

  var s = undefined;
  if (source[0].addEventListener && source[0].removeEventListener) {
    s = new _MulticastSource2.default(new EventTargetSource(event, source, useCapture));
  } else {
    throw new Error('source must support addEventListener/removeEventListener');
  }
  return new _Stream2.default(s);
};

exports.default = fromEvent;
