// most-subject/lib/Replay.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.replay = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _most = require('most');

var _multicast = require('@most/multicast');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function pushEvents(sink, buffer) {
  var i = 0;
  for (; i < buffer.length; ++i) {
    var item = buffer[i];
    sink.event(item.time, item.value);
  }
}

function replayAdd(sink) {
  var length = this._replayAdd(sink);
  if (this._replay.buffer.length > 0) {
    pushEvents(sink, this._replay.buffer);
  }
  return length;
}

function addToBuffer(event, replay) {
  if (replay.buffer.length >= replay.bufferSize) {
    replay.buffer.shift();
  }
  replay.buffer.push(event);
}

function replayEvent(time, value) {
  if (this._replay.bufferSize > 0) {
    addToBuffer({ time: time, value: value }, this._replay);
  }
  this._replayEvent(time, value);
}

var Replay = function () {
  function Replay(bufferSize, source) {
    _classCallCheck(this, Replay);

    this.source = source;
    this.bufferSize = bufferSize;
    this.buffer = [];
  }

  _createClass(Replay, [{
    key: 'run',
    value: function run(sink, scheduler) {
      if (sink._replay !== this) {
        sink._replay = this;
        sink._replayAdd = sink.add;
        sink.add = replayAdd;

        sink._replayEvent = sink.event;
        sink.event = replayEvent;
      }

      return this.source.run(sink, scheduler);
    }
  }]);

  return Replay;
}();

var replay = function replay(bufferSize, stream) {
  return new _most.Stream(new _multicast.MulticastSource(new Replay(bufferSize, stream.source)));
};

exports.replay = replay;
