// most-subject/lib/Subscription.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Subscription = function () {
  function Subscription() {
    var _this = this;

    _classCallCheck(this, Subscription);

    this.run = function (sink, scheduler) {
      return _this._run(sink, scheduler);
    };
    this.add = this.next = function (x) {
      return _this._add(x);
    };
    this.error = function (err) {
      return _this._error(err);
    };
    this.end = this.complete = function (x) {
      return _this._end(x);
    };
  }

  _createClass(Subscription, [{
    key: "_run",
    value: function _run(sink, scheduler) {
      this.sink = sink;
      this.scheduler = scheduler;
      this.active = true;
      return this;
    }
  }, {
    key: "dispose",
    value: function dispose() {
      this.active = false;
    }
  }, {
    key: "_add",
    value: function _add(x) {
      if (!this.active) {
        return;
      }
      tryEvent(this.sink, this.scheduler, x);
    }
  }, {
    key: "_error",
    value: function _error(e) {
      this.active = false;
      this.sink.error(this.scheduler.now(), e);
    }
  }, {
    key: "_end",
    value: function _end(x) {
      if (!this.active) {
        return;
      }
      this.active = false;
      tryEnd(this.sink, this.scheduler, x);
    }
  }]);

  return Subscription;
}();

function tryEvent(sink, scheduler, event) {
  try {
    sink.event(scheduler.now(), event);
  } catch (e) {
    sink.error(scheduler.now(), e);
  }
}

function tryEnd(sink, scheduler, event) {
  try {
    sink.end(scheduler.now(), event);
  } catch (e) {
    sink.error(scheduler.now(), e);
  }
}

exports.Subscription = Subscription;
