// cycle-snabbdom/lib/modules/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EventsModule = exports.HeroModule = exports.AttrsModule = exports.PropsModule = exports.ClassModule = exports.StyleModule = undefined;

var _class = require('snabbdom/modules/class');

var _class2 = _interopRequireDefault(_class);

var _props = require('snabbdom/modules/props');

var _props2 = _interopRequireDefault(_props);

var _attributes = require('snabbdom/modules/attributes');

var _attributes2 = _interopRequireDefault(_attributes);

var _eventlisteners = require('snabbdom/modules/eventlisteners');

var _eventlisteners2 = _interopRequireDefault(_eventlisteners);

var _styleModule = require('./style-module');

var _heroModule = require('./hero-module');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = [_styleModule.StyleModule, _class2.default, _props2.default, _attributes2.default];
exports.StyleModule = _styleModule.StyleModule;
exports.ClassModule = _class2.default;
exports.PropsModule = _props2.default;
exports.AttrsModule = _attributes2.default;
exports.HeroModule = _heroModule.HeroModule;
exports.EventsModule = _eventlisteners2.default;
